src/hotspot/share/memory/arena.cpp

```diff
@@ -68,50 +69,29 @@ class ChunkPool: public CHeapObj<mtInternal> {
     return c;
   }
 
- public:
-  // All chunks in a ChunkPool has the same size
-   ChunkPool(size_t size) : _size(size) { _first = NULL; _num_chunks = _num_used = 0; }
-
-  // Allocate a new chunk from the pool (might expand the pool)
-  NOINLINE void* allocate(size_t bytes, AllocFailType alloc_failmode) {
-    assert(bytes == _size, "bad size");
-    void* p = NULL;
-    // No VM lock can be taken inside ThreadCritical lock, so os::malloc
-    // should be done outside ThreadCritical lock due to NMT
-    { ThreadCritical tc;
-      _num_used++;
-      p = get_first();
-    }
-    if (p == NULL) p = os::malloc(bytes, mtChunk, CURRENT_PC);
-    if (p == NULL && alloc_failmode == AllocFailStrategy::EXIT_OOM) {
-      vm_exit_out_of_memory(bytes, OOM_MALLOC_ERROR, "ChunkPool::allocate");
-    }
-    return p;
-  }
-
   // Return a chunk to the pool
-  void free(Chunk* chunk) {
-    assert(chunk->length() + Chunk::aligned_overhead_size() == _size, "bad size");
+  void return_chunk(Chunk* chunk) {
+    assert(chunk->length() == _size, "wrong pool for this chunk");
     ThreadCritical tc;
-    _num_used--;
-
-    // Add chunk to list
     chunk->set_next(_first);
     _first = chunk;
     _num_chunks++;
   }
 
   // Prune the pool
-  void free_all_but(size_t n) {
+  void prune() {
+    static const int blocksToKeep = 5;
     Chunk* cur = NULL;
     Chunk* next;
     {
       // if we have more than n chunks, free all of them
       ThreadCritical tc;
-      if (_num_chunks > n) {
+      if (_num_chunks > blocksToKeep) {
         // free chunks at end of queue, for better locality
         cur = _first;
-        for (size_t i = 0; i < (n - 1) && cur != NULL; i++) cur = cur->next();
+        for (size_t i = 0; i < (blocksToKeep - 1) && cur != NULL; i++) {
+          cur = cur->next();
+        }
 
         if (cur != NULL) {
```

Comment:
[pre-existing] This test shouldn't be needed because `_num_chunks > blocksToKeep`.